public/app/core/components/FolderFilter/FolderFilter.tsx

```diff
@@ -0,0 +1,107 @@
+import React, { useCallback, useMemo, useState } from 'react';
+import { css } from '@emotion/css';
+import debounce from 'debounce-promise';
+import { AsyncMultiSelect, Icon, resetSelectStyles, useStyles2 } from '@grafana/ui';
+import { GrafanaTheme2, SelectableValue } from '@grafana/data';
+
+import { FolderInfo } from 'app/types';
+import { getBackendSrv } from 'app/core/services/backend_srv';
+
+export interface FolderFilterProps {
+  onChange: (folder: FolderInfo[]) => void;
+  maxMenuHeight?: number;
+}
+
+export function FolderFilter({ onChange: propsOnChange, maxMenuHeight }: FolderFilterProps): JSX.Element {
+  const styles = useStyles2(getStyles);
+  const [loading, setLoading] = useState(false);
+  const getOptions = useCallback((searchString: string) => getFoldersAsOptions(searchString, setLoading), []);
+  const debouncedLoadOptions = useMemo(() => debounce(getOptions, 300), [getOptions]);
+  const [value, setValue] = useState<Array<SelectableValue<FolderInfo>>>([]);
+  const onChange = useCallback(
+    (folders: Array<SelectableValue<FolderInfo>>) => {
+      const changedFolders = [];
+      for (const folder of folders) {
+        if (folder.value) {
+          changedFolders.push(folder.value);
+        }
+      }
+      propsOnChange(changedFolders);
+      setValue(folders);
+    },
+    [propsOnChange]
+  );
+  const selectOptions = {
+    defaultOptions: true,
+    isMulti: true,
+    noOptionsMessage: 'No folders found',
+    placeholder: 'Filter by folder',
+    styles: resetSelectStyles(),
+    maxMenuHeight,
+    value,
+    onChange,
+  };
+
+  return (
+    <div className={styles.container}>
+      {value.length > 0 && (
+        <span className={styles.clear} onClick={() => onChange([])}>
+          Clear folders
+        </span>
+      )}
+      <AsyncMultiSelect
+        {...selectOptions}
+        isClearable
+        isLoading={loading}
+        loadOptions={debouncedLoadOptions}
+        prefix={<Icon name="filter" />}
+        aria-label="Folder filter"
+      />
+    </div>
```

Comment:
I guess this is a conscious design decision:

![image](https://user-images.githubusercontent.com/2376619/116967534-ca1d3500-acb2-11eb-801c-91c5590ccff3.png)


I don't think we use this anywhere so far - are we trying to establish a new UX pattern? If so - do we still need the clear icon on the select? I would probably go with a single trigger for the same action to avoid unnecessary confusion.

Reply:
Wonder if we could reuse the tag filter from search here.
![Screenshot 2021-05-04 at 9 33 48](https://user-images.githubusercontent.com/8878045/116967831-e2458200-acbb-11eb-84fd-8e7abca9a75e.png)

Reply:
Not sure I follow. Do you mean the `Clear folders` part?. That "Clear xyz" is already used in a couple of pickers. FYI this folder filter is only used in Library Panel view.

Reply:
Yes, I was thinking about the Clear folders part. I think if we remove the clear icon from the select this should be just fine. I just don't like that the same action can be performed in two different ways :)

Reply:
Aha, I haven't noticed the clear icon in the select. Good catch, will look into this.

Reply:
Fixed